src/app_model/expressions/_expressions.py

```diff
@@ -182,22 +179,49 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         ast.fix_missing_locations(self)
         self._code = compile(ast.Expression(body=self), "<Expr>", "eval")
         self._names = set(_iter_names(self))
+        self.eval = self.eval_with_callables  # type: ignore
+
+    def eval(self, context: Mapping[str, object] | None = None) -> T:
+        """Evaluate this expression with names in `context`.
+
+        Parameters
+        ----------
+        context : Mapping[str, object] | None
+            Mapping of names to objects to evaluate the expression with.
+        """
+        # will have been replaced in __init__
+        raise NotImplementedError("This method should have been replaced.")
 
-    def eval(
-        self, context: Mapping[str, object] | None = None, **ctx_kwargs: object
-    ) -> T:
+    def eval_no_callables(self, context: Mapping[str, object] | None = None) -> T:
         """Evaluate this expression with names in `context`."""
         if context is None:
-            context = ctx_kwargs
-        elif ctx_kwargs:
-            context = {**context, **ctx_kwargs}
+            context = {}
         try:
-            return eval(self._code, {}, context)  # type: ignore
+            return eval(self._code, {}, context)  # type: ignore[no-any-return]
         except NameError as e:
-            miss = {k for k in self._names if k not in context}
-            raise NameError(
-                f"Names required to eval this expression are missing: {miss}"
-            ) from e
+            raise self._missing_names_error(context) from e
+
+    def eval_with_callables(self, context: Mapping[str, object] | None = None) -> T:
+        """Evaluate this expression with names in `context`, allowing callables."""
+        if context is None:
+            return self.eval_no_callables({})
+
+        # build a new context, evaluating any callables
+        # we only want to evaluate the callables if they are needed, so we
+        # build a new context with only the names in this expression.
+        ctx = {}
+        for name in self._names:
+            if name in context:
+                ctx[name] = val() if callable(val := context[name]) else val
+            else:
+                # early exit if we're missing names
+                raise self._missing_names_error(context)
```

Comment:
@tlambert03 I tried playing with stuff in `napari` and this branch, but I get this error with `napari` main when I right-click on a layer. Sometimes the error is thrown because `name = ''`, and sometimes the `name` makes sense but the `context` passed only contains the `app_model_context` keys. Not sure if it's something we're doing on our end that's been exposed by this branch, or if it's something on the app-model end.

Reply:
thanks for trying it out.  it's on the app-model end.  will write more later

Reply:
I tried and it works!